### PR TITLE
Contribute JIT write barrier documentation

### PR DESCRIPTION
**Pull request by 0xdaryl**
Describe the purpose and general operation of JIT write barriers for non-realtime
garbage collection policies.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
@Akira1Saitoh @knn-k FYI

**Comment by 0xdaryl** [comment]
@JamesKingdon

**Review by fjeremic — COMMENTED** [review]
Thanks for writing this!

File: doc/compiler/JitWriteBarriers.md [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+checks are required to determine that. Because the checks are performed inline
+the performance of the write barrier is greatly improved over calling into the
+GC each time to determine the correct course of action. The checks required
+depend on the GC policy and other features of the GC that are enabled.
+
+### Nullness Check
+
+Consider a reference store `A.field = B`. A write barrier is not required if `B`
+is provably null. This can be determined both at compile time (via the
+`isNonNull()` property on the source object node or at runtime with an explicit
```
Comment (dsouzai) [inline doc/compiler/JitWriteBarriers.md:100]: Needs a `)` after "source object node".

File: doc/compiler/JitWriteBarriers.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+         remember(A)
+      }
+   }
+```
+
+There are bits in an object's header that can be queried to determine whether
+the object resides in new space or old space.
+
+Typically, only the generational test is inlined and not the update to the
+remembered set. This is acceptable in practice becasue the remembered set update
```
Comment (dsouzai) [inline doc/compiler/JitWriteBarriers.md:131]: typo `becasue`

**Review by dsouzai — COMMENTED** [review]
Minor grammatical errors.

Thanks for writing this up Daryl.

**Comment by dsouzai** [comment]
Also worth adding `[ci skip]` to the commit message so that the travis build won't run as this is a doc only change.

File: doc/compiler/JitWriteBarriers.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+      /* Check to see if object is old.
+       * if ((object - vmThread->omrVMThread->heapBaseForBarrierRange0) < vmThread->omrVMThread->heapSizeForBarrierRange0) then old
+       *
+       * Since card dirtying also requires this calculation remember these values
+       */
+      UDATA base = (UDATA)vmThread->omrVMThread->heapBaseForBarrierRange0;
+      UDATA objectDelta = (UDATA)object - base;
+      UDATA size = vmThread->omrVMThread->heapSizeForBarrierRange0;
+      if (objectDelta < size) {
+         /* object is old so check for concurrent barrier */
```
Comment (Akira1Saitoh) [inline doc/compiler/JitWriteBarriers.md:305]: I've got a question. What is `old` object in non-generational GC policy? What exactly is `objectDelta < size` checking here?

File: doc/compiler/JitWriteBarriers.md [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+correctness.
+
+To do this efficiently, the heap is partitioned into a number of *cards* of a
+fixed size (e.g., 512 bytes) that are represented by a single byte in a *card
+table* structure. The mapping of an object address to a particular card is done
+with simple arithmetic and shifting on the object address. Rather than
+remembering individual objects to rescan, the card representing the memory where
+a destination object lives is "dirtied" at runtime (in practice, writing a `1`
+into the card table). All objects that reside in dirtied cards are rescanned on
+the next GC.
```
[the remark this answers is not in the document]
Comment (0xdaryl) [reply doc/compiler/JitWriteBarriers.md:151]: Yes, that's a mistake.  It's at the end of the current concurrent mark phase.  Thanks.

File: doc/compiler/JitWriteBarriers.md [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+In summary:
+
+| Write Barrier Kind | GC Policy and Description |
+|:----|:----|
+| j9gc_modron_wrtbar_none | `-Xgcpolicy:optthruput` : No write barrier is required |
+| j9gc_modron_wrtbar_always | A write barrier is always required.  The JIT always calls the GC write barrier helper for this case. |
+| j9gc_modron_wrtbar_cardmark | `-Xgcpolicy:optavgpause` : For non-generational GC policies, a concurrent mark check is required. |
+| j9gc_modron_wrtbar_cardmark_incremental | `-Xgcpolicy:balanced` : For region-based GCs, an inter-region barrier is required. |
+| j9gc_modron_wrtbar_cardmark_and_oldcheck | `-Xgcpolicy:gencon` : For generational GCs with concurrent mark, a generational check and a concurrent mark check is required. |
+| j9gc_modron_wrtbar_oldcheck | `-Xgcpolicy:gencon -Xconcurrentlevel0` : For generational GC policies, a generational check is required.  Concurrent mark is not being used. |
```
[the remark this answers is not in the document]
Comment (0xdaryl) [reply doc/compiler/JitWriteBarriers.md:62]: Perhaps, but I'm including it here for the convenience of the reader of this document and would rather not touch code as part of this PR unless something is wrong.

There are also metronome barriers in that list in j9modron.h that I don't touch on here (deliberately) and they should be part of any documentation effort in that file.

**Review by 0xdaryl — COMMENTED** [review]
Thank you all for your comments.  I believe I have addressed all of them.

I also added a link to a video that describes the mechanics of each of the GC policies in more detail if the reader needs more background.

**Review by fjeremic — APPROVED** [review]
My concerns have been addressed. Will wait for others who reviewed to give approvals and we should go ahead and merge this very useful piece of documentation. Thanks for writing this!

**Review by dsouzai — APPROVED** [review]

**Pull request merged**